docs/dev/test-architecture.adoc

```diff
@@ -238,10 +244,16 @@ Expect(output).To(ContainSubstring("No file changes detected, skipping build"))
 ----
 +
 
-* If oc, odo or generic library you are looking for is not present in helper package then create a new library function as per the scenario requirement.
+* If oc, odo or generic library you are looking for is not present in helper package then create a new library function as per the scenario requirement. 
++
+
+* Avoid logic implementation within the test files as it unnecessarily overloads the tests. Always create a new function with all the required logics within the helper package. Check for code duplicacy within helper and test files.
++
+
+* If you are looking for delay with a specific feature test, don't use `time.Sleep()` function, Instead use `timeout`.
```

Comment:
Ironically, we use this within our code
```sh
  1 tests/helper/helper_run.go|41 col 2| time.Sleep(timeout)                                                                                                                                                                                                                                                                                                                 
  2 tests/helper/helper_run.go|92 col 4| time.Sleep(time.Duration(intervalSeconds) * time.Second)
  3 || 
  4 tests/helper/helper_http.go|34 col 4| time.Sleep(time.Duration(interval) * time.Second)
  5 tests/helper/helper_http.go|45 col 3| time.Sleep(time.Duration(interval) * time.Second)
  6 || 
  7 tests/helper/helper_generic.go|140 col 3| time.Sleep(timeout)
  8 || 
  9 tests/integration/generic_test.go|349 col 4| time.Sleep(1 * time.Second)
```

But yeah, good tip, we shouldn't ever have time.sleep in the tests.

Reply:
> If you are looking for delay with a specific feature test, don't use `time.Sleep()` function, Instead use `timeout`.

@prietyc123 I am making the statement intuitive. PTAL 

If you are looking for delay with a specific feature test, don't use hard `time.Sleep()` function. Yes, you can use but as a polling interval of maximum duration. Check the helper package[link] for more such reference 

For example: 
```
func RetryInterval(maxRetry, intervalSeconds int, program string, args ...string) string {
	for i := 0; i < maxRetry; i++ {
		session := CmdRunner(program, args...)
		session.Wait()
		if session.ExitCode() == 0 {
			time.Sleep(time.Duration(intervalSeconds) * time.Second)
		} else {
			Consistently(session).ShouldNot(gexec.Exit(0), runningCmd(session.Command))
			return string(session.Err.Contents())
		}
	}
	Fail(fmt.Sprintf("Failed after %d retries", maxRetry))
	return ""
}
```
Or in Ginkgo there is a in build timeout feature [link] that you can make use of it.